File: Sources/PhotosStorage/S3Client.swift
import Foundation

/// A minimal S3 client for one bunny.net storage zone.
///
/// Concrete methods for exactly the operations the design needs — no generic
/// request layer. A `Sendable` struct holding immutable configuration, so callers
/// may use one instance from as many concurrent tasks as they like without an actor
/// serialising them.
///
/// Deliberately absent: a concurrency limiter. §9 measured that more concurrent
/// uploads makes throughput slightly *worse*, so that knob belongs in the caller's
/// worker pool where it can be seen, not buried here.
public struct S3Client: Sendable {

    public let storage: StorageURL
    public let signer: SigV4Signer
    public let transport: any HTTPTransport
    public let retry: RetryPolicy
    public let payloadSigning: PayloadSigning
    /// Uploads at or above this size go through multipart. Ordinary photos and
    /// shards never reach it; transcoded video does.
    public let multipartThreshold: Int64
    public let multipartPartSize: Int64

    public init(
        storage: StorageURL,
        secretAccessKey: String,
        transport: (any HTTPTransport)? = nil,
        retry: RetryPolicy = .default,
        payloadSigning: PayloadSigning = .signed,
        multipartThreshold: Int64 = 64 * 1024 * 1024,
        multipartPartSize: Int64 = 16 * 1024 * 1024
    ) {
        self.storage = storage
        // On bunny.net the zone name is the access key ID (§1).
        self.signer = SigV4Signer(
            accessKeyID: storage.zone,
            secretAccessKey: secretAccessKey,
            region: storage.region,
            service: "s3",
            normalizePath: false        // S3 signs the path exactly as sent
        )
        self.transport = transport ?? URLSessionTransport()
        self.retry = retry
        self.payloadSigning = payloadSigning
        self.multipartThreshold = multipartThreshold
        self.multipartPartSize = multipartPartSize
    }

    /// How the `x-amz-content-sha256` header is filled for file bodies.
    ///
    /// `.unsigned` avoids a full read pass over every uploaded byte, but it is only
    /// usable if bunny.net accepts `UNSIGNED-PAYLOAD` on header-authenticated PUTs —
    /// which a one-off manual probe settles. `.signed` is the choice that needs no
    /// verification, so it is the default.
    public enum PayloadSigning: Sendable {
        case signed
        case unsigned
    }

    // MARK: - Keys

    /// Builds the zone-relative request path for a key.
    ///
    /// NFC-normalises here, at the last gate before the wire: Linux stores whatever
    /// bytes it was handed and iOS emits NFD, so `Grün` composed and decomposed are
    /// different keys (§2). Doing it in the client makes the invariant structural
    /// rather than a rule every call site must remember.
    func path(forKey key: String) -> String {
        let normalized = key.precomposedStringWithCanonicalMapping
        let trimmed = normalized.hasPrefix("/") ? String(normalized.dropFirst()) : normalized
        return "/\(storage.zone)/\(trimmed)"
    }

    func url(path: String, query: [(name: String, value: String)]) -> URL? {
        var string = storage.endpoint.absoluteString
        if string.hasSuffix("/") { string.removeLast() }
        string += SigV4Signer.uriEncode(path, encodeSlash: false)
        if !query.isEmpty {
            string += "?" + SigV4Signer.canonicalQuery(query)
        }
        return URL(string: string)
    }

    // MARK: - Signing

    func signedRequest(
        method: String,
        path: String,
        query: [(name: String, value: String)] = [],
        headers: [(name: String, value: String)] = [],
        body: Body,
        payloadHash: String? = nil,
        now: Date = Date()
    ) throws -> HTTPRequest {
        guard let url = url(path: path, query: query) else {
            throw StorageError.malformedResponse("could not build a URL for \(path)")
        }

        let hash: String
        if let payloadHash {
            hash = payloadHash
        } else {
            switch body {
            case .empty:
                hash = SigV4Signer.sha256Hex(Data())
            case .data(let d):
                hash = SigV4Signer.sha256Hex(d)
            case .file(let fileURL):
                switch payloadSigning {
                case .unsigned: hash = SigV4Signer.unsignedPayload
                case .signed: hash = try SigV4Signer.sha256Hex(fileAt: fileURL)
                }
            }
        }

        var allHeaders: [(name: String, value: String)] = [
            ("Host", url.host ?? storage.endpoint.host ?? ""),
            ("X-Amz-Date", SigV4Signer.amzDate(now)),
            ("X-Amz-Content-Sha256", hash),
        ]
        allHeaders.append(contentsOf: headers)
        if let length = body.byteCount, length > 0 {
            allHeaders.append(("Content-Length", String(length)))
        }

        let canonical = signer.canonicalRequest(
            method: method, path: path, query: query,
            headers: allHeaders, payloadHash: hash
        )
        let sts = signer.stringToSign(canonicalRequest: canonical.text, date: now)
        let signature = signer.signature(stringToSign: sts, date: now)
        let credential = "\(signer.accessKeyID)/\(signer.credentialScope(date: now))"
        allHeaders.append((
            "Authorization",
            "\(SigV4Signer.algorithm) Credential=\(credential), "
                + "SignedHeaders=\(canonical.signedHeaders), Signature=\(signature)"
        ))

        return HTTPRequest(method: method, url: url, headers: allHeaders, body: body)
    }

    // MARK: - Sending, with retry

    /// Retries 5xx, 429 and transport failures only.
    ///
    /// 4xx is never retried — retrying a wrong password just delays the message §1
    /// promises to show. 412 is not a failure at all here; it is returned as a value
    /// because the caller must re-read the object before trying again.
    func send(_ request: HTTPRequest, expecting acceptable: Set<Int>,
              key: String? = nil) async throws -> HTTPResponse {
        var attempt = 0
        while true {
            attempt += 1
            do {
                let response = try await transport.send(request)
                if acceptable.contains(response.status) {
                    return response
                }
                let error = S3HTTPError.parse(
                    status: response.status,
                    body: response.body,
                    key: key,
                    requestID: response[header: "x-amz-request-id"]
                )
                guard retry.shouldRetry(status: response.status), attempt <= retry.maxAttempts else {
                    throw StorageError.http(error)
                }
                try await retry.wait(attempt: attempt,
                                     retryAfter: response[header: "retry-after"])
            } catch let error as StorageError {
                guard case .transport = error, attempt <= retry.maxAttempts else { throw error }
                try await retry.wait(attempt: attempt, retryAfter: nil)
            }
        }
    }

    // MARK: - Operations

    /// `nil` when the object does not exist.
    public func head(_ key: String) async throws -> S3Object? {
        let request = try signedRequest(method: "HEAD", path: path(forKey: key), body: .empty)
        let response = try await send(request, expecting: [200, 404], key: key)
        guard response.status == 200 else { return nil }
        return S3Object(
            key: key.precomposedStringWithCanonicalMapping,
            size: Int64(response[header: "content-length"] ?? "") ?? 0,
            etag: response[header: "etag"].map { ETag(header: $0) },
            lastModified: response[header: "last-modified"].flatMap(Self.parseHTTPDate)
        )
    }

    /// Fetches an object, optionally a byte range, optionally conditional.
    ///
    /// A 304 comes back as `.notModified` rather than an error: it is the common
    /// case in §4's sync diff, and burying the normal path inside a `catch` reads
    /// backwards and is easy to swallow.
    public func get(
        _ key: String,
        range: Range<Int64>? = nil,
        ifNoneMatch: ETag? = nil
    ) async throws -> GetResult {
        var headers: [(name: String, value: String)] = []
        if let range {
            headers.append(("Range", "bytes=\(range.lowerBound)-\(range.upperBound - 1)"))
        }
        if let ifNoneMatch {
            headers.append(("If-None-Match", ifNoneMatch.headerValue))
        }
        let request = try signedRequest(method: "GET", path: path(forKey: key),
                                        headers: headers, body: .empty)
        let response = try await send(request, expecting: [200, 206, 304], key: key)
        if response.status == 304 { return .notModified }
        return .object(response.body, response[header: "etag"].map { ETag(header: $0) })
    }

    /// Streams an object to a file. Used for previews and originals, which have no
    /// reason to pass through memory.
    @discardableResult
    public func download(
        _ key: String,
        to destination: URL,
        progress: (@Sendable (Int64, Int64) -> Void)? = nil
    ) async throws -> ETag? {
        var request = try signedRequest(method: "GET", path: path(forKey: key), body: .empty)
        request.downloadTo = destination
        request.progress = progress
        let response = try await send(request, expecting: [200], key: key)
        return response[header: "etag"].map { ETag(header: $0) }
    }

    /// Uploads an object, switching to multipart above `multipartThreshold`.
    ///
    /// `ifMatch` guards §2's single-owner shard rule: a stale ETag yields
    /// `.staleETag` (HTTP 412), meaning "someone else wrote it — re-read and retry".
    public func put(
        _ key: String,
        body: Body,
        contentType: String? = nil,
        ifMatch: ETag? = nil,
        progress: (@Sendable (Int64, Int64) -> Void)? = nil
    ) async throws -> PutResult {
        if case .file(let url) = body,
           let size = body.byteCount, size >= multipartThreshold, ifMatch == nil {
            let etag = try await multipartUpload(key, file: url, size: size,
                                                 contentType: contentType, progress: progress)
            return .written(etag)
        }

        var headers: [(name: String, value: String)] = []
        if let contentType { headers.append(("Content-Type", contentType)) }
        if let ifMatch { headers.append(("If-Match", ifMatch.headerValue)) }

        var request = try signedRequest(method: "PUT", path: path(forKey: key),
                                        headers: headers, body: body)
        request.progress = progress
        let response = try await send(request, expecting: [200, 412], key: key)
        if response.status == 412 { return .staleETag }
        return .written(response[header: "etag"].map { ETag(header: $0) })
    }

    /// Permanently removes an object. bunny.net has no versioning and no undelete,
    /// so this is the single irreversible operation in the system (§7). The iOS app
    /// never calls it.
    public func delete(_ key: String) async throws {
        let request = try signedRequest(method: "DELETE", path: path(forKey: key), body: .empty)
        _ = try await send(request, expecting: [200, 204, 404], key: key)
    }

    static func parseHTTPDate(_ s: String) -> Date? {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "GMT")
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        return formatter.date(from: s)
    }
}

File: domain/src/commonMain/kotlin/net/stho/photos/storage/S3Client.kt
package net.stho.photos.storage

import io.ktor.client.HttpClient
import io.ktor.client.content.ProgressListener
import io.ktor.client.request.header
import io.ktor.client.plugins.onDownload
import io.ktor.client.plugins.onUpload
import io.ktor.client.request.prepareRequest
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsBytes
import io.ktor.client.statement.bodyAsChannel
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.content.OutgoingContent
import io.ktor.utils.io.ByteWriteChannel
import io.ktor.utils.io.readAvailable
import io.ktor.utils.io.writeFully
import kotlin.time.Clock
import kotlin.time.Instant
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toInstant
import kotlinx.io.buffered
import kotlinx.io.files.Path
import kotlinx.io.files.SystemFileSystem
import kotlinx.io.write
import net.stho.photos.S3HttpFailure

/**
 * A minimal S3 client for one bunny.net storage zone.
 *
 * Concrete methods for exactly the operations the design needs — no generic request layer. The
 * configuration is immutable and the Ktor client underneath is safe to share, so callers may
 * use one instance from as many concurrent coroutines as they like.
 *
 * The [http] client is supplied rather than built here: Ktor *is* the transport abstraction
 * (§7), which makes its engine the seam a test replaces and its retry plugin the whole retry
 * policy — see [retryStorageFailures].
 *
 * Deliberately absent: a concurrency limiter. §9 measured that more concurrent uploads makes
 * throughput slightly *worse*, so that knob belongs in the caller's worker pool where it can be
 * seen, not buried here.
 */
public class S3Client(
    public val storage: StorageUrl,
    secretAccessKey: String,
    private val http: HttpClient,
    public val payloadSigning: PayloadSigning = PayloadSigning.UNSIGNED,
    /**
     * Uploads at or above this size go through multipart. Ordinary photos and shards never
     * reach it; transcoded video does.
     */
    public val multipartThreshold: Long = 64L * 1024 * 1024,
    public val multipartPartSize: Long = 16L * 1024 * 1024,
    internal val clock: Clock = Clock.System,
) {

    /** On bunny.net the zone name is the access key ID (§1). */
    internal val signer: SigV4Signer = SigV4Signer(
        accessKeyId = storage.zone,
        secretAccessKey = secretAccessKey,
        region = storage.region,
        service = "s3",
        // S3 signs the path exactly as sent.
        normalizePath = false,
    )

    /**
     * How the `x-amz-content-sha256` header is filled for file bodies.
     *
     * **Verified against the live zone: bunny.net accepts `UNSIGNED-PAYLOAD` on
     * header-authenticated PUTs**, so that is the default — it saves a full read pass over every
     * uploaded byte, which across ~34 000 objects is a second pass over 120 GB competing with
     * the encoders.
     *
     * In-memory bodies are always hashed for real; they are small, and the pass costs nothing.
     * [SIGNED] remains available for a service that rejects it.
     */
    public enum class PayloadSigning { SIGNED, UNSIGNED }

    // ------------------------------------------------------------------ operations

    /** `null` when the object does not exist. */
    public suspend fun head(key: String): S3Object? {
        val response = send("HEAD", keyPath(key), accepting = setOf(200, 404), key = key)
        if (response.status.value != 200) return null
        return S3Object(
            key = key,
            size = response.headers[HttpHeaders.ContentLength]?.toLongOrNull() ?: 0,
            etag = response.headers[HttpHeaders.ETag]?.asETag(),
            lastModified = response.headers[HttpHeaders.LastModified]?.asHttpDate(),
        )
    }

    /**
     * Fetches an object, optionally a byte range, optionally conditional.
     *
     * A 304 comes back as [GetResult.NotModified] rather than a failure: it is the common case
     * in §4's sync diff, and burying the normal path inside a `catch` reads backwards and is
     * easy to swallow.
     */
    public suspend fun get(
        key: String,
        range: LongRange? = null,
        ifNoneMatch: ETag? = null,
    ): GetResult {
        val headers = buildList {
            // Inclusive at both ends, which is what the header means and what a `LongRange` is.
            if (range != null) add(HttpHeaders.Range to "bytes=${range.first}-${range.last}")
            if (ifNoneMatch != null) add(HttpHeaders.IfNoneMatch to ifNoneMatch.headerValue)
        }
        val response = send(
            "GET", keyPath(key), headers = headers,
            accepting = setOf(200, 206, 304), key = key,
        )
        if (response.status.value == 304) return GetResult.NotModified
        return GetResult.Content(response.bodyAsBytes(), response.headers[HttpHeaders.ETag]?.asETag())
    }

    /**
     * Streams an object to a file. Used for previews and originals, which have no reason to
     * pass through memory.
     */
    public suspend fun download(key: String, to: Path, progress: ProgressListener? = null): ETag? {
        val request = sign("GET", keyPath(key))
        return reachingZone {
            http.prepareRequest(request.url) {
                method = HttpMethod.Get
                request.headers.forEach { (name, value) -> header(name, value) }
                if (progress != null) onDownload(progress)
            }.execute { response ->
                if (response.status.value != 200) throw response.asFailure(key)
                SystemFileSystem.sink(to).buffered().use { sink ->
                    val channel = response.bodyAsChannel()
                    val chunk = ByteArray(1 shl 16)
                    while (true) {
                        val read = channel.readAvailable(chunk, 0, chunk.size)
                        if (read <= 0) break
                        sink.write(chunk, 0, read)
                    }
                }
                response.headers[HttpHeaders.ETag]?.asETag()
            }
        }
    }

    /**
     * Uploads an object, switching to multipart above [multipartThreshold].
     *
     * [ifMatch] guards §2's single-owner shard rule: a stale ETag yields [PutResult.StaleETag]
     * (HTTP 412), meaning "someone else wrote it — re-read and retry".
     */
    public suspend fun put(
        key: String,
        body: Body,
        contentType: String? = null,
        ifMatch: ETag? = null,
        progress: ProgressListener? = null,
    ): PutResult {
        val size = body.byteCount
        if (body is Body.File && size != null && size >= multipartThreshold && ifMatch == null) {
            return PutResult.Written(multipartUpload(key, body.path, size, contentType, progress))
        }

        val headers = buildList {
            if (contentType != null) add(HttpHeaders.ContentType to contentType)
            if (ifMatch != null) add(HttpHeaders.IfMatch to ifMatch.headerValue)
        }
        val response = send(
            "PUT", keyPath(key), headers = headers, body = body,
            accepting = setOf(200, 412), key = key, progress = progress,
        )
        if (response.status.value == 412) return PutResult.StaleETag
        return PutResult.Written(response.headers[HttpHeaders.ETag]?.asETag())
    }

    /**
     * Permanently removes an object. bunny.net has no versioning and no undelete, so this is
     * the single irreversible operation in the system (§7). The iOS app never calls it.
     */
    public suspend fun delete(key: String) {
        send("DELETE", keyPath(key), accepting = setOf(200, 204, 404), key = key)
    }

    // ------------------------------------------------------------------ keys and URLs

    /**
     * Builds the zone-relative request path for a key. Path-style, because bunny.net addresses
     * a zone as the first path segment rather than as a subdomain.
     */
    internal fun keyPath(key: String): String = "/${storage.zone}/${key.removePrefix("/")}"

    internal fun requestUrl(path: String, query: List<Pair<String, String>>): String = buildString {
        append(storage.endpoint)
        append(path.uriEncoded(encodeSlash = false))
        if (query.isNotEmpty()) append('?').append(query.canonicalQuery())
    }

    // ------------------------------------------------------------------ signing and sending

    /** A signed request, ready for the wire: URL, the headers to attach, and the payload. */
    internal class SignedRequest(
        val url: String,
        val headers: List<Pair<String, String>>,
        val body: Body,
    )

    internal fun sign(
        method: String,
        path: String,
        query: List<Pair<String, String>> = emptyList(),
        headers: List<Pair<String, String>> = emptyList(),
        body: Body = Body.Empty,
    ): SignedRequest {
        val now = clock.now()
        val hash = when (body) {
            Body.Empty -> EMPTY_PAYLOAD_SHA256
            is Body.Bytes -> body.value.sha256Hex()
            is Body.File -> when (payloadSigning) {
                PayloadSigning.UNSIGNED -> UNSIGNED_PAYLOAD
                PayloadSigning.SIGNED -> body.path.sha256Hex()
            }
        }

        // Signed but not attached: the HTTP client derives `Host` from the URL and
        // `Content-Length` from the body, and a second copy of either is rejected or sent twice.
        // The values are the ones it will send, so the signature still matches.
        val derived = buildList {
            add(HttpHeaders.Host to storage.host)
            body.byteCount?.takeIf { it > 0 }?.let { add(HttpHeaders.ContentLength to it.toString()) }
        }
        val attached = buildList {
            add("X-Amz-Date" to now.amzDate())
            add("X-Amz-Content-Sha256" to hash)
            addAll(headers)
        }

        val canonical = signer.canonicalRequest(
            method = method,
            path = path,
            query = query,
            headers = derived + attached,
            payloadHash = hash,
        )
        val signature = signer.signature(signer.stringToSign(canonical.text, now), now)
        val authorization = "$AWS4_HMAC_SHA256 Credential=${signer.credential(now)}, " +
            "SignedHeaders=${canonical.signedHeaders}, Signature=$signature"

        return SignedRequest(
            url = requestUrl(path, query),
            headers = attached + (HttpHeaders.Authorization to authorization),
            body = body,
        )
    }

    /**
     * Signs, sends, and turns an unacceptable status into the failure §1 promises to show.
     *
     * Retry is not here: it is Ktor's `HttpRequestRetry` plugin on [http], which has already
     * exhausted its attempts by the time a response reaches this line.
     */
    internal suspend fun send(
        method: String,
        path: String,
        query: List<Pair<String, String>> = emptyList(),
        headers: List<Pair<String, String>> = emptyList(),
        body: Body = Body.Empty,
        accepting: Set<Int>,
        key: String? = null,
        progress: ProgressListener? = null,
    ): HttpResponse {
        val request = sign(method, path, query, headers, body)
        val response = reachingZone {
            http.request(request.url) {
                this.method = HttpMethod.parse(method)
                request.headers.forEach { (name, value) -> header(name, value) }
                when (val payload = request.body) {
                    Body.Empty -> Unit
                    is Body.Bytes -> setBody(payload.value)
                    is Body.File -> setBody(FileContent(payload.path))
                }
                if (progress != null) onUpload(progress)
            }
        }
        if (response.status.value !in accepting) throw response.asFailure(key)
        return response
    }

    private suspend fun HttpResponse.asFailure(key: String?): S3HttpFailure =
        bodyAsBytes().asS3HttpFailure(status.value, key, headers[AMZ_REQUEST_ID])
}

private const val AMZ_REQUEST_ID = "x-amz-request-id"

private val EMPTY_PAYLOAD_SHA256 = ByteArray(0).sha256Hex()

/**
 * Request payloads. [File] exists because §8's background uploads must come from a file on
 * disk, and because a 200 MB transcode should never sit in memory.
 */
public sealed interface Body {
    public data object Empty : Body

    public data class Bytes(public val value: ByteArray) : Body {
        override fun equals(other: Any?): Boolean =
            this === other || (other is Bytes && value.contentEquals(other.value))

        override fun hashCode(): Int = value.contentHashCode()
    }

    public data class File(public val path: Path) : Body
}

/** Null when a file body's size cannot be read — the file is gone, or is not a file. */
public val Body.byteCount: Long?
    get() = when (this) {
        Body.Empty -> 0
        is Body.Bytes -> value.size.toLong()
        is Body.File -> SystemFileSystem.metadataOrNull(path)?.size
    }

/**
 * A file body, streamed rather than read into memory.
 *
 * [contentLength] is declared rather than left to chunked encoding on purpose: SigV4 signs
 * `Content-Length`, so a chunked upload would send no header for one that was signed and every
 * request would come back 403.
 */
private class FileContent(private val path: Path) : OutgoingContent.WriteChannelContent() {
    override val contentLength: Long? = SystemFileSystem.metadataOrNull(path)?.size

    override suspend fun writeTo(channel: ByteWriteChannel) {
        SystemFileSystem.source(path).buffered().use { source ->
            val chunk = ByteArray(1 shl 16)
            while (true) {
                val read = source.readAtMostTo(chunk, 0, chunk.size)
                if (read <= 0) break
                channel.writeFully(chunk, 0, read)
            }
        }
        channel.flushAndClose()
    }
}

private val HTTP_MONTHS =
    listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

/**
 * `Wed, 15 Jan 2026 10:30:00 GMT` — RFC 7231's one mandatory date format.
 *
 * Hand-parsed for the same reason EXIF dates are: the format is fixed and its zone is always
 * GMT, so a formatter would only add locale behaviour that has to be pinned anyway. Null when
 * the header is anything else, since a missing modification date costs nothing.
 */
internal fun String.asHttpDate(): Instant? {
    val parts = substringAfter(',').trim().split(' ').filter { it.isNotEmpty() }
    if (parts.size < 4) return null
    val day = parts[0].toIntOrNull() ?: return null
    val month = HTTP_MONTHS.indexOf(parts[1]) + 1
    if (month == 0) return null
    val year = parts[2].toIntOrNull() ?: return null
    val time = parts[3].split(':').mapNotNull(String::toIntOrNull)
    if (time.size != 3) return null
    return runCatching {
        LocalDateTime(year, month, day, time[0], time[1], time[2]).toInstant(TimeZone.UTC)
    }.getOrNull()
}
